lib/kms/wrapper.js

```diff
@@ -10,49 +10,106 @@ const KMIPClient = require('arsenal').network.kmipClient;
-} catch (error) {
-    logger.warn('scality kms unavailable. ' +
-      'Using file kms backend unless mem specified.',
-      { error });
-    scalityKMS = file;
-    scalityKMSImpl = 'fileKms';
+const { fromArn, KmsProtocol } = require('arsenal/build/lib/network/KMSInterface');
+
+function getScalityKms() {
+    let scalityKMS;
+    let scalityKMSImpl;
+
+    try {
+        // eslint-disable-next-line import/no-unresolved
+        const ScalityKMS = require('scality-kms');
+        // TODO add scality arn in scality-kms
+        scalityKMS = new ScalityKMS(config.kms);
+        scalityKMSImpl = 'scalityKms';
+    } catch (error) {
+        logger.warn('scality kms unavailable. ' +
+        'Using file kms backend unless mem specified.',
+        { error });
+        scalityKMS = file;
+        scalityKMSImpl = 'fileKms';
+    }
+    return { scalityKMS, scalityKMSImpl };
+}
+
+function getClient(kms) {
+    let client;
+    let implName;
+
+    if (kms === 'mem') {
+        client = inMemory;
+        implName = 'memoryKms';
+    } else if (kms === 'file' || kms === 'cdmi') {
+        client = file;
+        implName = 'fileKms';
+    } else if (kms === 'scality') {
+        ({ scalityKMS: client, scalityKMSImpl: implName } = getScalityKms());
+    } else if (kms === 'kmip') {
+        const kmipConfig = { kmip: config.kmip };
+        if (!kmipConfig.kmip) {
+            throw new Error('KMIP KMS driver configuration is missing.');
+        }
+        client = new KMIPClient(kmipConfig);
+        implName = 'kmip';
+    } else if (kms === 'aws') {
+        const awsConfig = { kmsAWS: config.kmsAWS };
+        client = new KmsAWSClient(awsConfig);
+        implName = 'aws';
+    } else {
+        throw new Error(`KMS backend is not configured: ${kms}`);
+    }
+
+    return { client, implName };
 }
 
-let client;
-let implName;
-
-if (config.backends.kms === 'mem') {
-    client = inMemory;
-    implName = 'memoryKms';
-} else if (config.backends.kms === 'file' || config.backends.kms === 'cdmi') {
-    client = file;
-    implName = 'fileKms';
-} else if (config.backends.kms === 'scality') {
-    client = scalityKMS;
-    implName = scalityKMSImpl;
-} else if (config.backends.kms === 'kmip') {
-    const kmipConfig = { kmip: config.kmip };
-    if (!kmipConfig.kmip) {
-        throw new Error('KMIP KMS driver configuration is missing.');
+/**
+ * Note: non current instance from previous keys won't be healthchecked
+ * `{ [arnPrefix]: clientDetails }`
+ */
+const clientInstances = {};
+
+const { client, implName } = getClient(config.backends.kms);
+const { type, protocol, provider } = fromArn(client.arnPrefix);
+clientInstances[`${type}:${protocol}:${provider}`] = { client, implName };
+
+const mapKmsProtocolToClient = {
+    [KmsProtocol.aws_kms]: 'aws',
+    // others already match
+};
+
+// TODO store KMS identifier differently to avoid concatenations + split of arn
+// multiple times for a request
+function getClientForKey(key) {
```

Comment:
Can we be certain that key always follows the new format?

Reply:
We should be, but I need to validate the format of the key passed to  PutObject if a scality arn is provided. The other places it's generated internally so we are good

Reply:
I've added a validation function if the key matches the scality arn format.

Now we are sure the format is valid before using it with the KMS